### PR TITLE
Fix local and remote cover coverage and build and push pyrsia docker …

### LINKED ISSUE #245: Automate release process to release dockerized version of Pyrsia
Process to release the dockerized version
to a dockerhub organization with proper name and version attached to it

PyrsiaOSS orgname
image name pyrsia-node

Include a github action to automate this.

### PR DESCRIPTION
**Pull request by sbtaylor15**
Resolves #245, #230

* Moved Dockerfile to installers/docker to be consistent with the other installers
* Replaced code coverage steps with generic action.  
* Added `docker build` for the Base Code Coverage image to speed up local code coverage runs.
* Added `docker push` for Base Code Coverage to DockerHub.
* Added `docker build` for the pyrsia_node based on the apt-get.
* Added `docker push` for the pyrsia_node image to DockHub.  Use `docker pull pyrsiaos/pyrsia-node` to fetch the image.  To run:  `docker run -it -p 7888:7888 pyrsiaos/pyrsia-node`.  Logging is set to stdout and stderr.
* Updated `tests/test_code_coverage.sh` to use the Base Code Coverage image + local source code to create and run a local image. 
 
### Setup needed on Forked Repo and https://codecov.io for Code Coverage

You will need to setup a https://codecov.io free account and associate your fork to that account.  This will provide Code Coverage Reports in your fork prior to a PR.  Create a Repository Secret _CODECOV_TOKEN_ on your fork that is codecov.io token.  This will enable the action to upload the reports.  See images below:

image

image

image

### CODE REVIEW

File: installers/docker/DockerfileCodeCoverage [hunk 1]
```diff
@@ -5,6 +5,7 @@ RUN rustup --version; \
     cargo --version; \
     rustc --version; \
     cargo install cargo-tarpaulin;
-COPY . /home/pyrsia/
+
 WORKDIR /home/pyrsia
-CMD ["cargo", "tarpaulin", "--workspace"]
+
+ENTRYPOINT ["cargo", "tarpaulin", "--workspace"]
```
Comment (prince-chrismc) [inline installers/docker/DockerfileCodeCoverage:11]: It would be nice to use the `.Dockerfile` as the extension since GitHub is able to do syntax highlighting 

You can see the difference on these lines 
image

File: .github/workflows/rust.yml [hunk 2]
```diff
@@ -164,11 +166,45 @@ jobs:
...
+    - name: Only build if DockerfileCodeCoverage changed
+      uses: tj-actions/changed-files@v13.1
+      id: changed-files
+
+    - name: Build Code Coverage Base Image
+      if: contains(steps.changed-files.outputs.all_changed_and_modified_files, 'DockerfileCodeCoverage')
+      run: | 
+        docker build -t pyrsiaos/codecoverage:1.0 -f installers/docker/DockerfileCodeCoverage .
+        docker login --username ${{ secrets.DOCKERHUB_USERID }} --password '${{ secrets.DOCKERHUB_PASSWORD }}'
+        docker push pyrsiaos/codecoverage:1.0
```
Comment (prince-chrismc) [inline .github/workflows/rust.yml:204]: There are official actions from the Docker team for building and pushing images, did you consider using those instead?

https://github.com/marketplace/actions/build-and-push-docker-images

File: .github/workflows/rust.yml [hunk 3]
```diff
@@ -154,6 +154,8 @@ jobs:
     runs-on: ubuntu-latest
     steps:
     - uses: actions/checkout@v2
+      with:
+        fetch-depth: 0
```
Comment (prince-chrismc) [inline .github/workflows/rust.yml:158]: Whats the reason for getting all the history? it seems only the event commit is being used but I am not familar with all the new actions being used

https://github.com/marketplace/actions/checkout says

> Set `fetch-depth: 0` to fetch all history for all branches and tags.

Just curious why 😄

File: installers/docker/DockerfileAptGet [hunk 4]
```diff
@@ -0,0 +1,12 @@
+FROM ubuntu:bionic
+
+EXPOSE 7888
```
Comment (prince-chrismc) [inline installers/docker/DockerfileAptGet:3]: is this blocked by #263 ?

File: .github/workflows/rust.yml [hunk 5]
```diff
@@ -164,11 +166,45 @@ jobs:
...
+        override: true
+
+    - name: Run cargo-tarpaulin
+      uses: actions-rs/tarpaulin@v0.1
+      with:
+        args: '-- --test-threads 1'
 
-    - uses: codecov/codecov-action@v2
+    - name: Upload to codecov.io
+      uses: codecov/codecov-action@v1.0.2
```
Comment (prince-chrismc) [inline .github/workflows/rust.yml:181]: why is this downgraded from `v2`? 🤔

File: .github/workflows/rust.yml [hunk 6]
```diff
@@ -164,11 +166,45 @@ jobs:
...
+    - name: Install stable toolchain
+      uses: actions-rs/toolchain@v1
+      with:
+        toolchain: stable
+        override: true
+
+    - name: Run cargo-tarpaulin
+      uses: actions-rs/tarpaulin@v0.1
+      with:
+        args: '-- --test-threads 1'
```
Comment (prince-chrismc) [inline .github/workflows/rust.yml:178]: I've been trying to avoid this action because it does very little only set the same default output format https://github.com/actions-rs/tarpaulin/blob/044a1e5bdace8dd2f727b1af63c1d9a1d3572068/src/main.ts#L27 

Second, the maintainer is MIA, their GitHub status... 

image

I've been trying to reach them for weeks so we can contribute to the other `action-rs` actions we are using but I've had no luck

File: .github/workflows/rust.yml (hunk 5 above)
Comment (prince-chrismc) [reply .github/workflows/rust.yml:181]: I think the v1 uses the old bash uploader 🤔 

https://github.com/marketplace/actions/codecov#%EF%B8%8F--deprecration-of-v1 points to
 https://about.codecov.io/security-update/

**Comment by prince-chrismc** [comment]
Talking on Zoom, most of the changes are from copy pasting of the https://github.com/actions-rs/tarpaulin#example-workflow

**Review by mgrand — DISMISSED** [review]

**Review by prince-chrismc — APPROVED** [review]
It's finally working!!!! 🚀 

Amazing work!

**Review by mgrand — DISMISSED** [review]

**Review by mgrand — APPROVED** [review]

File: .github/workflows/rust.yml [hunk 7]
```diff
@@ -164,11 +166,55 @@ jobs:
...
+        context: .
+        push: true
+        file: installers/docker/AptGet.Dockerfile
+        tags: pyrsiaoss/pyrsia-node:latest
+
+    - name: Only build if CodeCoverage.Dockerfile changed
+      uses: tj-actions/changed-files@v13.1
+      id: changed-files
+
+    - name: Build and Push Code Coverage Base Image
```
Comment (erwin1) [inline .github/workflows/rust.yml:213]: little question: wouldn't it make more sense to split the docker job: one part that does the docker image building and push and one part that does the code coverage. They don't seem related other than they both push to docker hub?

Comment (prince-chrismc) [reply .github/workflows/rust.yml:213]: This is build the code coverage image to save time. It's not the "prysia image".

Steve did some optimization here... you can see the next line there's an `if` so it's only done when required

**Review by prince-chrismc — APPROVED** [review]
